thinc/backends/cupy_ops.py

```diff
@@ -75,11 +75,23 @@ def asarray(self, data, dtype=None):
         dtype = {"dtype": dtype} if dtype is not None else {}
         if isinstance(data, cupy.ndarray):
             return self.xp.asarray(data, **dtype)
-        elif hasattr(data, "data_ptr"):
+        elif (
+            hasattr(data, "data_ptr")
+            and hasattr(data, "storage")
+            and hasattr(data, "device")
+        ):
             # Handles PyTorch Tensors
-            pointer = cupy.cuda.MemoryPointer(data.data_ptr())
-            shape = data.stride()
-            array = self.xp.ndarray(shape, memptr=pointer, **dtype)
+            assert data.device.type == "cuda", "Non-CUDA tensor passed to CupyOps"
+
+            ext_pointer_wrapper = cupy.cuda.UnownedMemory(
+                ptr=data.data_ptr(),
+                size=data.storage().size(),
+                owner=data,
+                device_id=data.device.index,
+            )
+
+            pointer = cupy.cuda.MemoryPointer(ext_pointer_wrapper, offset=0)
+            array = self.xp.ndarray(shape=data.shape, memptr=pointer, **dtype)
```

Comment:
The semantics of `dtype` are incorrect. If `dtype` is not specified, the data type of the original array must be used, otherwise:

```
>>> x = torch.tensor([42.0], device='cuda')
>>> y = ops.asarray(x)
>>> y
array([5.48372203e-315])
```

If `dtype` is specified and differs from the `dtype` of the original array, a copy should be made (converting to the specified `dtype`).